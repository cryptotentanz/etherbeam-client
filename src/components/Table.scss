@use '../styles/resources/sizes';
@use '../styles/resources/colors';

.my-table {
  border-collapse: collapse;
  border-style: solid;
  border-width: sizes.$border;
  width: 100%;
  padding: sizes.$spacer;

  th {
    letter-spacing: sizes.$letter-spacing;
  }

  th,
  td {
    border-style: solid;
    border-width: sizes.$border;
    padding: sizes.$spacer-s;
  }

  // &-primary {
  //   background-color: colors.$bg-primary;
  //   border-color: colors.$bd-primary;
  //   color: colors.$fg-light;

  // }

  // &-secondary {
  //   background-color: colors.$bg-secondary;
  //   border-color: colors.$bd-secondary;
  //   color: colors.$fg-light;

  //   th,
  //   td {
  //     border-color: colors.$bd-secondary;
  //   }
  // }

  // &-dark {
  //   background-color: colors.$bg-dark;
  //   border-color: colors.$bd-dark;
  //   color: colors.$fg-dark;

  //   th,
  //   td {
  //     border-color: colors.$bd-dark;
  //   }
  // }

  // &-light {
  //   background-color: colors.$bg-light;
  //   border-color: colors.$bd-light;
  //   color: colors.$fg-light;

  //   th,
  //   td {
  //     border-color: colors.$bd-light;
  //   }
  // }

  // &-positive {
  //   background-color: colors.$bg-positive;
  //   border-color: colors.$bd-positive;
  //   color: colors.$fg-dark;

  //   th,
  //   td {
  //     border-color: colors.$bd-positive;
  //   }
  // }

  // &-negative {
  //   background-color: colors.$bg-negative;
  //   border-color: colors.$bd-negative;
  //   color: colors.$fg-dark;

  //   th,
  //   td {
  //     border-color: colors.$bd-negative;
  //   }
  // }
}
