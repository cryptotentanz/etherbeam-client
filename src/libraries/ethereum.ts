import { Address, AddressType } from './types'

const ETHERSCAN_URL_BASE = 'https://etherscan.io/'

export type EtherscanUrlType = 'block' | 'tx' | 'token' | 'address'

export const getEtherscanUrl = (type: EtherscanUrlType, id: string | number): string => {
  return `${ETHERSCAN_URL_BASE}${type}/${id}`
}

export const createTransactionAddress = (hash: string): Address => {
  return {
    hash,
    type: AddressType.Transaction,
    url: getEtherscanUrl('tx', hash),
  }
}
