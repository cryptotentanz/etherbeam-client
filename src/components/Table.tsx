import React, { FC, ReactNode } from 'react'
import './Table.scss'

type ColorProp = 'dark' | 'light' | 'primary' | 'secondary' | 'positive' | 'negative'

interface TableProps {
  color?: ColorProp
  children?: ReactNode
}

const Table: FC<TableProps> = ({ color = 'dark', children }: TableProps) => {
  return <table className={`my-table my-${color}-bg`}>{children}</table>
}

export default Table
