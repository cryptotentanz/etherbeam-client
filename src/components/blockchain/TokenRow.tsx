import React, { FC } from 'react'
import { Token } from 'libraries/types'
import TableRow from '../TableRow'
import AddressItem from './AddressItem'
import TokenPriceHistory from './TokenPriceHistory'
import TokenAmount from './TokenAmount'

interface TokenRowProps {
  token: Token
  goToTokenPage: (hash: string) => void
}

const TokenRow: FC<TokenRowProps> = ({ token, goToTokenPage }: TokenRowProps) => {
  const { hash, name, price, priceHistory } = token

  return (
    <TableRow action={() => goToTokenPage(hash)} clickable>
      <td>{name}</td>
      <td>
        <TokenAmount amount={price} noBadge noSymbol />
        {!!priceHistory?.length && (
          <TokenPriceHistory size="s" className="my-tokenrow-pricehistory" priceHistory={priceHistory} />
        )}
      </td>
      <td className="my-d-min-xs my-text-center">
        <AddressItem address={token} short noFlag />
      </td>
    </TableRow>
  )
}

export default TokenRow
