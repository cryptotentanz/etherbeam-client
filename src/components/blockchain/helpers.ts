import { faCoins, faFile, faWallet, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import { BigNumber, ethers } from 'ethers'
import numeral from 'numeral'
import { AddressType, Token } from 'libraries/types'

const ETH_DECIMALS = 18
const MAX_DECIMALS = 13
const UNISWAP_SWAP_URL = 'https://app.uniswap.org/#/swap'

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
}

const printAmount = (amount: string | null, decimals: number, options?: AmountPrintOptions | null): string => {
  const { forceDecimals, decimals: forcedDecimals } = options ?? {}
  const decimalsCount = Math.min(forcedDecimals || decimals, MAX_DECIMALS)
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`

  return amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber | null, options?: AmountPrintOptions | null, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_DECIMALS, options)

export const printTokenAmount = (
  amount: BigNumber | null,
  token: Token,
  options?: AmountPrintOptions | null
): string => {
  const { decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition | null => {
  switch (type) {
    case AddressType.Wallet:
      return faWallet
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faCoins
    default:
      return null
  }
}

export const getUniswapTradeUrl = (inputHash: string | null, outputHash: string | null): string => {
  return `${UNISWAP_SWAP_URL}?${inputHash ? `inputCurrency=${inputHash}&` : ''}${
    outputHash ? `outputCurrency=${outputHash}` : ''
  }`
}
