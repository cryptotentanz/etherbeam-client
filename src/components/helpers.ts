import { UserType } from 'libraries/types'
import { faUser, faUserShield, IconDefinition } from '@fortawesome/free-solid-svg-icons'

export const getUserTypeIcon = (type: UserType): IconDefinition => {
  switch (type) {
    case UserType.Administrator:
      return faUserShield
    default:
      return faUser
  }
}
