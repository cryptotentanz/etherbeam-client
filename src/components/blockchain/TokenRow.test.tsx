import * as React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import { act } from 'react-dom/test-utils'
import { clearMocks } from 'tests/helpers'
import { createToken } from 'tests/fixtures/ethereum'
import Table from '../Table'
import TokenRow from './TokenRow'

describe('<TokenRow />', () => {
  const token = createToken('0x0000000000000000000000000000000000000111')
  const goToTokenPage = jest.fn()

  let component: ReactWrapper

  describe('Default', () => {
    beforeAll(() => {
      component = mount(
        <Table>
          <tbody>
            <TokenRow token={token} goToTokenPage={goToTokenPage} />
          </tbody>
        </Table>
      )
    })

    it('should render amount', () => expect(component.find('TokenAmount').prop('amount')).toBe(token.price))

    it('should render price history', () =>
      expect(component.find('TokenPriceHistory').prop('priceHistory')).toBe(token.priceHistory))

    it('should render address item', () => expect(component.find('AddressItem').prop('address')).toBe(token))

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('On click', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(
        <Table>
          <tbody>
            <TokenRow token={token} goToTokenPage={goToTokenPage} />
          </tbody>
        </Table>
      )

      act(() => {
        component.find('TableRow').simulate('click')
        component.update()
      })
    })

    it('should call action', () => {
      expect(goToTokenPage).toBeCalledTimes(1)
      expect(goToTokenPage.mock.calls[0][0]).toBe(token.hash)
    })
  })
})
