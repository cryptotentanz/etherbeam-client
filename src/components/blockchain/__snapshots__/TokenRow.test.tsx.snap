// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokenRow /> Default should match snapshot 1`] = `
<table
  class="my-table my-dark-bg"
>
  <tbody>
    <tr
      class="my-dark-bg my-clickable"
      tabindex="0"
    >
      <td>
        Token
      </td>
      <td>
        <span>
          <span
            class="my-tokenamount-amount"
          >
             NaN
          </span>
        </span>
        <ul
          class="my-list-h  my-tokenrow-pricehistory"
        >
          <li
            class="my-listitem "
          >
            <span
              class="my-badge my-positive-bg my-size-s "
            >
              <span
                class="my-difference-prefix"
              >
                H
              </span>
              <span
                class="my-difference-amount"
              >
                +100%
              </span>
            </span>
          </li>
          <li
            class="my-listitem "
          >
            <span
              class="my-badge my-negative-bg my-size-s "
            >
              <span
                class="my-difference-prefix"
              >
                D
              </span>
              <span
                class="my-difference-amount"
              >
                -50%
              </span>
            </span>
          </li>
          <li
            class="my-listitem "
          >
            <span
              class="my-badge my-negative-bg my-size-s "
            >
              <span
                class="my-difference-prefix"
              >
                W
              </span>
              <span
                class="my-difference-amount"
              >
                -75%
              </span>
            </span>
          </li>
          <li
            class="my-listitem "
          >
            <span
              class="my-badge my-light-bg my-size-s "
            >
              <span
                class="my-difference-prefix"
              >
                M
              </span>
              <span
                class="my-difference-amount"
              >
                +0%
              </span>
            </span>
          </li>
          <li
            class="my-listitem "
          >
            <span
              class="my-badge my-positive-bg my-size-s "
            >
              <span
                class="my-difference-prefix"
              >
                Y
              </span>
              <span
                class="my-difference-amount"
              >
                +300%
              </span>
            </span>
          </li>
        </ul>
      </td>
      <td
        class="my-d-min-xs my-text-center"
      >
        <span>
          <a
            class="my-clickable  my-link my-secondary-fg"
            href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
            rel="noreferrer"
            tabindex="0"
            target="_blank"
          >
            <span
              class="my-addresshash"
            >
              0x0000...0111
            </span>
          </a>
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;
